MennaMohie/assignment2/Solution.cpp

```diff
@@ -12,118 +14,111 @@ class Node
-            return true;
-        return (contains(currentRoot->left, val) || contains(currentRoot->right, val));
-    }
-
-    bool resetRoot(T val)
+    int Size;
+    add(Node<T>* currentRoot, T parentVal, T val, direction dir)
     {
-        if(root==nullptr)
-        {
-            root=new Node<T>(val);
-            return true;
-        }
-        else if(!contains(root, val))
-        {
-            root->value=val;
-            return true;
-        }
-        else
-            return false;
-    }
-
-    //adds a new node (val) as a left child to an existing node (parentVal)
-    bool addAsLeftChild(Node<T> *currentRoot, T parentVal, T val)
-    {
-        if(!contains(root, parentVal) || contains(root, val) || currentRoot==nullptr)
-            return false;
         if(currentRoot->value==parentVal)
         {
-            if(currentRoot->left==nullptr)
+            if(dir==0)
             {
-                currentRoot->left=new Node<T>(val);
-                return true;
+                if(currentRoot->left==nullptr)
+                {
+                    currentRoot->left=new Node<T>(val);
+                    Size++;
+                    return true;
+                }
+                else
+                    return false;
             }
-            else
+            if(dir==1)
             {
-                return false;
+                if(currentRoot->right==nullptr)
+                {
+                    currentRoot->right=new Node<T>(val);
+                    return true;
+                }
+                else
+                    return false;
             }
         }
-        if(addAsLeftChild(currentRoot->left, parentVal, val))
-            return true;
-        return addAsLeftChild(currentRoot->right, parentVal, val);
+        if(dir==0)
+        {
+            if(addChildToParent(currentRoot->left, parentVal, val, addLeft))
+                return true;
+            return addChildToParent(currentRoot->right, parentVal, val, addLeft);
+        }
+        if(dir==1)
+        {
+            if(addChildToParent(currentRoot->left, parentVal, val, addRight))
+                return true;
+            return addChildToParent(currentRoot->right, parentVal, val, addRight);
+        }
+    }
+public:
+    BinaryTree(T val):root(new Node<T>(val)),Size(1) {}
+
+    Node<T>* getRoot()
+    {
+        return root;
+    }
 
+    Node<T>* contains(Node<T> *currentRoot, T val)
+    {
+        if(currentRoot==nullptr || currentRoot->value==val)
+            return currentRoot;
+        Node<T> *leftSide = contains(currentRoot->left, val);
+        Node<T> *rightSide = contains(currentRoot->right, val);
+        return (leftSide==nullptr?rightSide:leftSide);
     }
 
-    bool addAsRightChild(Node<T> *currentRoot, T parentVal, T val)
+    //adds a new node (val) as a left/right child (depends on dir) to an existing node (parentVal)
+    bool addChildToParent(Node<T> *currentRoot, T parentVal, T val, direction dir)
```

Comment:
You don't need the currentRoot parameter, just pass root to the add method bellow.